File: 0317/트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 트리 {
	
	static int[] parent;
	static boolean[] isCycle; // 해당 루트 노드가 사이클을 포함하는지

	
	public static int find(int x) {
	    if (parent[x] == x) return x; // 자기 자신이 부모면 루트 노드
        return parent[x] = find(parent[x]); // 경로 압축 적용
	}
	
	public static void union(int a, int b) {
	    int rootA = find(a);
	    int rootB = find(b);

	   // 이미 같은 루트 -> 사이클 발생
       if (rootA == rootB) {
        isCycle[rootA] = true;
        } 
        else {
        // 작은 루트 번호를 부모로 설정
	    	if(rootA < rootB) {
	    		parent[rootB] = rootA;
	    		if (isCycle[rootB]) isCycle[rootA] = true; // 사이클 정보 전달
	    	}
	    	else {
	    		parent[rootA] = rootB;
	    		if (isCycle[rootA]) isCycle[rootB] = true;
	    	}
	    }
	}
	
	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder  sb = new StringBuilder ();
		int caseNum = 1;
		
		while(true) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			
			int n = Integer.parseInt(st.nextToken());
			int m = Integer.parseInt(st.nextToken());
			
			if (n==0 && m==0) break;
			
			parent = new int[n+1];
			isCycle = new boolean[n+1];
			
			for (int i = 1; i <= n; i++) {
				parent[i] = i; // 각 노드는 자기 자신을 부모로 설정
			}

			for (int j = 0; j < m; j++) {
				st = new StringTokenizer(br.readLine());
				
				int u = Integer.parseInt(st.nextToken());
				int v = Integer.parseInt(st.nextToken());
				union(u, v);
			}
			
			int treeCnt = 0;
			for (int k = 1; k <=n; k++) {
				if(parent[k] == k && !isCycle[k]) { // 자기 자신이 루트이며 사이클이 없는 경우
					treeCnt++;
				}
			}
			
			if (treeCnt == 0) {
                sb.append("Case ").append(caseNum).append(": No trees.\n");
            } else if (treeCnt == 1) {
                sb.append("Case ").append(caseNum).append(": There is one tree.\n");
            } else {
                sb.append("Case ").append(caseNum).append(": A forest of ").append(treeCnt).append(" trees.\n");
            }
			
			caseNum++;
		}
		System.out.println(sb);
	}
}

File: 0317/트리.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 9)

def dfs(v):
    global k
    
    if not visited[v]:
        visited[v] = 1
        result[v] = k
        
        for ele in graph[v]:
            if not visited[ele]:
                dfs(ele)

def check_cycle(v, c):
    global check
    
    for ele in graph[v]:
        if ele != c:
            if ele in visit_v:
                check = 1
                return
            visit_v.append(ele)
            check_cycle(ele, v)
        else:
            continue

case_num = 1
        
while True:
    n, m = map(int, input().split())
    
    if n == 0 and m == 0:
        break
    
    graph = [[] for _ in range(n + 1)]
    visited = [1] + [0] * n
    
    for _ in range(m):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    result = [0] * (n + 1)
    k = 1
    is_cycle = 0

    for i in range(1, n + 1):
        dfs(i)
        k += 1

    list_arr = []

    for i in range(1, n + 1):
        tmp = []
        for j in range(1, n + 1):
            if result[j] == i:
                tmp.append(j)
        if tmp:
            list_arr.append(tmp)

    for ele in list_arr:
        visit_v = []
        check = 0
        check_cycle(ele[0], 0)
        if check == 1:
                is_cycle += 1

    tmp = len(set(result)) - is_cycle - 1

    print(f'Case {case_num}: ', end = '')
    case_num += 1

    if tmp == 1:
        print('There is one tree.')
    elif tmp > 1:
        print(f'A forest of {tmp} trees.')
    else:
        print('No trees.')
